File: core/web/src/main/java/edu/uci/ics/texera/web/resource/UserFileResource.java
package edu.uci.ics.texera.web.resource;

import static org.jooq.impl.DSL.defaultValue;

import java.io.InputStream;
import java.nio.file.Paths;
import java.sql.Connection;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import javax.servlet.http.HttpSession;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

import org.apache.commons.lang3.tuple.Pair;
import org.glassfish.jersey.media.multipart.FormDataContentDisposition;
import org.glassfish.jersey.media.multipart.FormDataParam;
import org.jooq.DSLContext;
import org.jooq.Record1;
import org.jooq.Record5;
import org.jooq.Result;
import org.jooq.types.UInteger;

import static edu.uci.ics.texera.dataflow.jooq.generated.Tables.*;

import edu.uci.ics.texera.web.TexeraWebException;
import edu.uci.ics.texera.dataflow.resource.file.FileManager;
import edu.uci.ics.texera.dataflow.sqlServerInfo.UserSqlServer;
import edu.uci.ics.texera.web.response.GenericWebResponse;
import io.dropwizard.jersey.sessions.Session;

@Path("/user/file")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
public class UserFileResource {
    
    /**
     * Corresponds to `src/app/dashboard/type/user-file.ts`
     */
    public static class UserFile {
        public UInteger id; // the ID in MySQL database is unsigned int
        public String name;
        public String path;
        public String description;
        public UInteger size; // the size in MySQL database is unsigned int

        public UserFile(UInteger id, String name, String path, String description, UInteger size) {
            this.id = id;
            this.name = name;
            this.path = path;
            this.description = description;
            this.size = size;
        }
    }
        
    /**
     * This method will handle the request to upload a single file.
     * 
     * @param uploadedInputStream
     * @param fileDetail
     * @return
     */
    @POST
    @Path("/upload")
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    public GenericWebResponse uploadFile(
            @FormDataParam("file") InputStream uploadedInputStream,
            @FormDataParam("file") FormDataContentDisposition fileDetail,
            @FormDataParam("size") String size,
            @FormDataParam("description") String description,
            @Session HttpSession session
            ) {
        UInteger userID = UserResource.getUserFromSession(session).getUserID();
        String fileName = fileDetail.getFileName();
        UInteger sizeUInteger = parseStringToUInteger(size);
        
        Pair<Boolean, String> validationResult = validateFileName(fileName, userID);
        if (!validationResult.getLeft()) {
            return new GenericWebResponse(1, validationResult.getRight());
        }
        
        this.handleFileUpload(uploadedInputStream, fileName, description, sizeUInteger, userID);
        return new GenericWebResponse(0, "success");
    }
    
    @GET
    @Path("/list")
    public List<UserFile> listUserFiles(@Session HttpSession session){
        UInteger userID = UserResource.getUserFromSession(session).getUserID();
        
        Result<Record5<UInteger, String, String, String, UInteger>> result = getUserFileRecord(userID);
        
        if (result == null) return new ArrayList<>();
        
        List<UserFile> fileList = result.stream()
                .map(
                    record -> new UserFile(
                            record.get(USERFILE.FILEID),
                            record.get(USERFILE.NAME),
                            record.get(USERFILE.PATH),
                            record.get(USERFILE.DESCRIPTION),
                            record.get(USERFILE.SIZE)
                            )
                        ).collect(Collectors.toList());
        
        return fileList;
    }
    
    @DELETE
    @Path("/delete/{fileID}")
    public GenericWebResponse deleteUserFile(@PathParam("fileID") String fileID, @Session HttpSession session) {
        UInteger userID = UserResource.getUserFromSession(session).getUserID();
        UInteger fileIdUInteger = parseStringToUInteger(fileID);
        Record1<String> result = deleteInDatabase(fileIdUInteger, userID);
        
        if (result == null) return new GenericWebResponse(1, "The file does not exist");
        
        String filePath = result.get(USERFILE.PATH);
        FileManager.getInstance().deleteFile(Paths.get(filePath));
        
        return new GenericWebResponse(0, "success");
    }
    
    @POST
    @Path("/validate")
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    public GenericWebResponse validateUserFile(@Session HttpSession session, @FormDataParam("name") String fileName) {
        UInteger userID = UserResource.getUserFromSession(session).getUserID();
        Pair<Boolean, String> validationResult = validateFileName(fileName, userID);
        return new GenericWebResponse(validationResult.getLeft() ? 0 : 1, validationResult.getRight());
    }
    
    private Record1<String> deleteInDatabase(UInteger fileID, UInteger userID) {
        // Connection is AutoCloseable so it will automatically close when it finishes.
        try (Connection conn = UserSqlServer.getConnection()) {
            DSLContext create = UserSqlServer.createDSLContext(conn);
            
            /**
             * Known problem for jooq 3.x
             * delete...returning clause does not work properly
             * retrieve the filepath first, then delete it.
             */
            Record1<String> result = create
                    .select(USERFILE.PATH)
                    .from(USERFILE)
                    .where(USERFILE.FILEID.eq(fileID).and(USERFILE.USERID.equal(userID)))
                    .fetchOne();
            
            int count = create
                    .delete(USERFILE)
                    .where(USERFILE.FILEID.eq(fileID).and(USERFILE.USERID.equal(userID)))
                    //.returning(USERFILE.FILEPATH) does not work
                    .execute();
            
            throwErrorWhenNotOne("delete file " + fileID + " failed in database", count);
            
            return result;
            
        } catch (Exception e) {
            throw new TexeraWebException(e);
        }
    }
    
    private Result<Record5<UInteger, String, String, String, UInteger>> getUserFileRecord(UInteger userID) {
        // Connection is AutoCloseable so it will automatically close when it finishes.
        try (Connection conn = UserSqlServer.getConnection()) {
            DSLContext create = UserSqlServer.createDSLContext(conn);
            
            Result<Record5<UInteger, String, String, String, UInteger>> result = create
                    .select(USERFILE.FILEID, USERFILE.NAME, USERFILE.PATH, USERFILE.DESCRIPTION, USERFILE.SIZE)
                    .from(USERFILE)
                    .where(USERFILE.USERID.equal(userID))
                    .fetch();
            
            return result;
            
        } catch (Exception e) {
            throw new TexeraWebException(e);
        }
    }
    
    private void handleFileUpload(InputStream fileStream, String fileName, String description, UInteger size, UInteger userID) {
        int count = insertFileToDataBase(
                fileName, 
                FileManager.getFilePath(userID.toString(), fileName).toString(),
                size,
                description,
                userID);
        
        throwErrorWhenNotOne("Error occurred while inserting file record to database", count);
        
        FileManager.getInstance().storeFile(fileStream, fileName, userID.toString());
    }
    
    private UInteger parseStringToUInteger(String userID) throws TexeraWebException {
        try {
            return UInteger.valueOf(userID);
        } catch (NumberFormatException e) {
            throw new TexeraWebException("Incorrect String to Double");
        }
    }
    
    
    private int insertFileToDataBase(String fileName, String path, UInteger size, String description, UInteger userID) {
        // Connection is AutoCloseable so it will automatically close when it finishes.
        try (Connection conn = UserSqlServer.getConnection()) {
            DSLContext create = UserSqlServer.createDSLContext(conn);
            
            int count = create.insertInto(USERFILE)
                    .set(USERFILE.USERID,userID)
                    .set(USERFILE.FILEID, defaultValue(USERFILE.FILEID))
                    .set(USERFILE.NAME, fileName)
                    .set(USERFILE.PATH, path)
                    .set(USERFILE.DESCRIPTION, description)
                    .set(USERFILE.SIZE, size)
                    .execute();
            
            return count;
            
        } catch (Exception e) {
            throw new TexeraWebException(e);
        }
    }
    
    private Pair<Boolean, String> validateFileName(String fileName, UInteger userID) {
        if (fileName == null) {
            return Pair.of(false, "file name cannot be null");
        } else if (fileName.trim().isEmpty()) {
            return Pair.of(false, "file name cannot be empty");
        } else if (isFileNameExisted(fileName, userID)){
            return Pair.of(false, "file name already exists");
        } else {
            return Pair.of(true, "filename validation success");
        }
    }
    
    private Boolean isFileNameExisted(String fileName, UInteger userID) {
        try (Connection conn = UserSqlServer.getConnection()) {
            DSLContext create = UserSqlServer.createDSLContext(conn);
            
            boolean result = create
                    .fetchExists(
                            create.selectFrom(USERFILE)
                            .where(USERFILE.USERID.equal(userID)
                                    .and(USERFILE.NAME.equal(fileName)))
                            );
            return result;
        } catch (Exception e) {
            throw new TexeraWebException(e);
        }
    }
    
    /**
     * Most the sql operation should only be executed once. eg. insertion, deletion.
     * this method will raise TexeraWebException when the input number is not one
     * @param errorMessage the message displaying when raising the error
     * @param number the number to be checked with one
     * @throws TexeraWebException
     */
    private void throwErrorWhenNotOne(String errorMessage, int number) throws TexeraWebException {
        if (number != 1) {
            throw new TexeraWebException(errorMessage);
        }
    }
}

File: core/amber/src/main/scala/edu/uci/ics/texera/web/resource/dashboard/user/file/UserFileResource.scala
package edu.uci.ics.texera.web.resource.dashboard.user.file

import edu.uci.ics.texera.Utils
import edu.uci.ics.texera.web.SqlServer
import edu.uci.ics.texera.web.auth.SessionUser
import edu.uci.ics.texera.web.model.jooq.generated.Tables._
import edu.uci.ics.texera.web.model.jooq.generated.enums.UserFileAccessPrivilege
import edu.uci.ics.texera.web.model.jooq.generated.tables.daos.{FileDao, UserFileAccessDao}
import edu.uci.ics.texera.web.model.jooq.generated.tables.pojos.{File, User, UserFileAccess}
import edu.uci.ics.texera.web.resource.dashboard.user.file.UserFileResource.{
  DashboardFile,
  context,
  fileDao,
  saveFile
}
import io.dropwizard.auth.Auth
import org.apache.commons.lang3.tuple.Pair
import org.glassfish.jersey.media.multipart.FormDataParam
import org.jooq.DSLContext
import org.jooq.types.UInteger

import java.io.{InputStream, OutputStream}
import java.net.URLDecoder
import java.nio.file.{Files, Paths}
import java.util
import java.util.UUID
import javax.annotation.security.RolesAllowed
import javax.ws.rs._
import javax.ws.rs.core.{MediaType, Response, StreamingOutput}
import scala.collection.JavaConverters._
import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

object UserFileResource {
  final private lazy val context: DSLContext = SqlServer.createDSLContext
  final private lazy val fileDao = new FileDao(context.configuration)
  final private lazy val userFileAccessDao = new UserFileAccessDao(context.configuration)

  def saveFile(uid: UInteger, fileName: String, stream: InputStream, des: String = ""): Unit = {
    val path = Utils.amberHomePath.resolve("user-resources").resolve("files").resolve(uid.toString)
    Files.createDirectories(path)
    val filepath = path.resolve(UUID.randomUUID.toString)
    Files.copy(stream, filepath)
    val file = new File(
      uid,
      null,
      UInteger.valueOf(filepath.toFile.length()),
      fileName,
      filepath.toString,
      des,
      null
    )
    fileDao.insert(file)
    userFileAccessDao.merge(
      new UserFileAccess(
        uid,
        file.getFid,
        UserFileAccessPrivilege.WRITE
      )
    )
  }

  case class DashboardFile(
      ownerEmail: String,
      accessLevel: String,
      file: File
  )
}
@Produces(Array(MediaType.APPLICATION_JSON))
@RolesAllowed(Array("REGULAR", "ADMIN"))
@Path("/user/file")
class UserFileResource {
  @POST
  @Consumes(Array(MediaType.MULTIPART_FORM_DATA))
  @Path("/upload")
  def uploadFile(
      @FormDataParam("file") stream: InputStream,
      @FormDataParam("name") fileName: String,
      @Auth user: SessionUser
  ): Response = {
    val uid = user.getUid
    val validationResult = validateFileName(fileName, uid)
    if (!validationResult.getLeft) {
      return Response
        .status(Response.Status.BAD_REQUEST)
        .entity(validationResult.getRight)
        .build()
    }
    saveFile(uid: UInteger, fileName, stream)
    Response.ok().build()
  }

  @GET
  @Path("/list")
  def getFileList(@Auth sessionUser: SessionUser): util.List[DashboardFile] = {
    getFileRecord(sessionUser.getUser)
  }

  private def getFileRecord(user: User): util.List[DashboardFile] = {
    val fids: mutable.ArrayBuffer[UInteger] = mutable.ArrayBuffer()
    val fileEntries: mutable.ArrayBuffer[DashboardFile] = mutable.ArrayBuffer()
    context
      .select()
      .from(USER_FILE_ACCESS)
      .join(FILE)
      .on(USER_FILE_ACCESS.FID.eq(FILE.FID))
      .join(USER)
      .on(FILE.OWNER_UID.eq(USER.UID))
      .where(USER_FILE_ACCESS.UID.eq(user.getUid))
      .fetch()
      .forEach(fileRecord => {
        fids += fileRecord.into(FILE).getFid
        fileEntries += DashboardFile(
          fileRecord.into(USER).getEmail,
          fileRecord.into(USER_FILE_ACCESS).getPrivilege.toString,
          fileRecord.into(FILE).into(classOf[File])
        )
      })

    context
      .select()
      .from(FILE_OF_WORKFLOW)
      .join(FILE)
      .on(FILE_OF_WORKFLOW.FID.eq(FILE.FID))
      .join(USER)
      .on(FILE.OWNER_UID.eq(USER.UID))
      .join(WORKFLOW_USER_ACCESS)
      .on(FILE_OF_WORKFLOW.WID.eq(WORKFLOW_USER_ACCESS.WID))
      .where(WORKFLOW_USER_ACCESS.UID.eq(user.getUid))
      .fetch()
      .forEach(fileRecord => {
        if (!fileEntries.exists(file => { file.file.getFid == fileRecord.into(FILE).getFid })) {
          fileEntries += DashboardFile(
            fileRecord.into(USER).getEmail,
            "READ",
            fileRecord.into(FILE).into(classOf[File])
          )
        }
      })
    fileEntries.toList.asJava
  }

  @GET
  @Path("/autocomplete/{query:.*}")
  def autocompleteUserFiles(
      @Auth sessionUser: SessionUser,
      @PathParam("query") q: String
  ): util.List[String] = {
    // get the user files
    // select the filenames that applies the input
    val query = URLDecoder.decode(q, "UTF-8")
    val user = sessionUser.getUser
    val fileList: List[DashboardFile] = getFileRecord(user).asScala.toList
    val filenames = ArrayBuffer[String]()
    val username = user.getEmail
    // get all the filename list
    for (i <- fileList) {
      filenames += i.file.getName
    }
    // select the filenames that apply
    val selectedByFile = ArrayBuffer[String]()
    val selectedByUsername = ArrayBuffer[String]()
    val selectedByFullPath = ArrayBuffer[String]()
    for (e <- filenames) {
      val fullPath = username + "/" + e
      if (e.contains(query) || query.isEmpty)
        selectedByFile += (username + "/" + e)
      else if (username.contains(query))
        selectedByUsername += (username + "/" + e)
      else if (fullPath.contains(query))
        selectedByFullPath += (username + "/" + e)
    }
    (selectedByFile ++ selectedByUsername ++ selectedByFullPath).toList.asJava
  }

  @DELETE
  @Path("/delete/{fid}")
  def deleteFile(
      @PathParam("fid") fid: UInteger,
      @Auth user: SessionUser
  ): Unit = {
    Files.deleteIfExists(Paths.get(fileDao.fetchOneByFid(fid).getPath))
    fileDao.deleteById(fid)
  }

  @GET
  @Path("/download/{fid}")
  def downloadFile(
      @PathParam("fid") fid: UInteger,
      @Auth user: SessionUser
  ): Response = {
    Response
      .ok(
        new StreamingOutput() {
          @Override
          def write(output: OutputStream): Unit = {
            Files.copy(Paths.get(fileDao.fetchOneByFid(fid).getPath), output)
            output.flush()
          }
        },
        MediaType.APPLICATION_OCTET_STREAM
      )
      .header(
        "content-disposition",
        String.format("attachment; filename=%s", fileDao.fetchOneByFid(fid).getName)
      )
      .build
  }

  @PUT
  @Path("/name/{fid}/{name}")
  def changeFileName(
      @PathParam("fid") fid: UInteger,
      @PathParam("name") name: String,
      @Auth user: SessionUser
  ): Unit = {
    val validationRes = this.validateFileName(name, user.getUid)
    if (!validationRes.getLeft) {
      throw new BadRequestException(validationRes.getRight)
    } else {
      val userFile = fileDao.fetchOneByFid(fid)
      userFile.setName(name)
      fileDao.update(userFile)
    }
  }

  @PUT
  @Path("/description/{fid}/{description}")
  def changeFileDescription(
      @PathParam("fid") fid: UInteger,
      @PathParam("description") description: String
  ): Unit = {
    val userFile = fileDao.fetchOneByFid(fid)
    userFile.setDescription(description)
    fileDao.update(userFile)
  }

  private def validateFileName(fileName: String, userID: UInteger): Pair[Boolean, String] = {
    if (
      context.fetchExists(
        context
          .selectFrom(FILE)
          .where(FILE.OWNER_UID.equal(userID).and(FILE.NAME.equal(fileName)))
      )
    ) Pair.of(false, "file name already exists")
    else Pair.of(true, "filename validation success")
  }

}
